mlir/lib/Bindings/Python/PybindUtils.h

```diff
@@ -231,7 +231,7 @@ class Sliceable {
   /// Returns the element at the given slice index. Supports negative indices
   /// by taking elements in inverse order. Returns a nullptr object if out
   /// of bounds.
-  pybind11::object getItem(intptr_t index) {
+  virtual pybind11::object getItem(intptr_t index) {
```

Comment:
This isn't for RTTI, this is purely so that `OpResultList` and `OpBlockArgumentList` can do something different from e.g. what `PyAffineMapExprList` does. Namely `OpResultList::getItem` and `BlockArgumentList::getItem` call `item.attr(MLIR_PYTHON_MAYBE_DOWNCAST_ATTR)()`.

Reply:
This class is CRTP, so it should be possible to do something like

```
pybind11::object getItem(intptr_t index) {
  index = wrapIndex(index);
  // other fluff
  return static_cast<Derived *>(this)->getItemImpl(index);
}

pybind11::object getItemImpl(intptr_t index) {
  // default implementation lives here
}
```

and have some subclasses provide different implementations of `getItemImpl`.

I also suspect it is possible to avoid the C++/Python back-and-forth here. The current implementation calls `getRawElement` that returns an instance of `ElementTy`, casts it to `py::object` and hands it off to the override that calls `maybeDownCast` via a python attribute. It should be possible to just directly call `maybeDownCast` on an instance of `ElementTy` since we have it. Potentially with some `llvm::is_detected` templating around the element class.

Reply:
all set (using the detection idiom you recommended)